design/open/openapi-multipart-upload.md

```diff
@@ -0,0 +1,217 @@
+# OpenAPI for S3 Like Multipart Upload Proposal
+
+## Introduction
+
+This proposal outlines the development of Open API endpoints to enable multipart upload capabilities similar to Amazon S3 multipart. The intention is to enhance the data uploading process for large files, ensuring efficient and reliable data transfer.
+
+## Goal
+
+- Implementing multi-part upload APIs with support for parallel upload optimizes data transfer for large files over varying network conditions using lakectl.
+
+## Non-goals
+
+- Support all underlying storage types as we can experiment with the one that currently support this feature.
+- SDK to wrap the new API and support this capability.
+
+
+## Features and Functionalities
+
+In order to minimaize the scope of this feature and understand the benefits before supporting all blockstores, this feature will be scoped to S3 blockstore and only if presigned capability is enabled.
+
+Initiation of Upload: The API will allow clients to initiate a multipart upload session, assigning a unique upload ID for subsequent operations. The client will require pass the number of parts it requires and the call will provide a set of URLs to upload each part.
```

Comment:
Agree, any streaming or working with unknown size is good for multipart or at least working with "upload manager" that will perform multipart or single object upload.

In order to limit the scope of the implementation, it is a limitation as I didn't want the client to call lakeFS for each part.
Working with local file and known in advance the size we like to upload - optional list as part of the create to start multipart will handle it on a single request.
The suggested API doesn't require the server to return any presigned part so we can include this capability later without breaking the API.

Reply:
OK, so perhaps document next possible step -- it isn't clear?

For instance somewhere...
```md
### Next steps

None of the returned URLs has to be used, it is fine to ask for more than are needed.  In future we may add an _additional_ API call to URLs for uploading more parts.  This will allow more "streaming" uses, for instance as parallels to how Hadoop S3A uses the S3 MPU API.  The current API is compatible with such an addition.
```